src/PowerShellEditorServices/Services/Analysis/PssaCmdletAnalysisEngine.cs

```diff
@@ -0,0 +1,548 @@
+                        pssaModuleInfo = ps.Invoke<PSModuleInfo>()?
+                            .OrderByDescending(moduleInfo => moduleInfo.Version)
+                            .FirstOrDefault();
+                    }
+                }
+                catch (Exception e)
+                {
+                    throw new FileNotFoundException("Unable to find PSScriptAnalyzer module on the module path", e);
+                }
+
+                if (pssaModuleInfo == null)
+                {
+                    throw new FileNotFoundException("Unable to find PSScriptAnalyzer module on the module path");
+                }
+
+                // Now that we know where the PSScriptAnalyzer we want to use is,
+                // create a base session state with PSScriptAnalyzer loaded
+#if DEBUG
+                InitialSessionState sessionState = Environment.GetEnvironmentVariable("PSES_TEST_USE_CREATE_DEFAULT") == "1"
+                    ? InitialSessionState.CreateDefault()
+                    : InitialSessionState.CreateDefault2();
+#else
+                InitialSessionState sessionState = InitialSessionState.CreateDefault2();
+#endif
+
+                sessionState.ImportPSModule(new [] { pssaModuleInfo.ModuleBase });
+
+                RunspacePool runspacePool = RunspaceFactory.CreateRunspacePool(sessionState);
+
+                runspacePool.SetMaxRunspaces(1);
+                runspacePool.ThreadOptions = PSThreadOptions.ReuseThread;
+
+                // Open the runspace pool here so we can deterministically handle the PSModulePath change issue
+                using (PSModulePathPreserver.Take())
+                {
+                    runspacePool.Open();
+                }
+
+                return runspacePool;
+            }
+        }
+
+        /// <summary>
+        /// Wraps the result of an execution of PowerShell to send back through
+        /// asynchronous calls.
+        /// </summary>
+        private class PowerShellResult
+        {
+            public PowerShellResult(
+                Collection<PSObject> output,
+                PSDataCollection<ErrorRecord> errors,
+                bool hasErrors)
+            {
+                Output = output;
+                Errors = errors;
+                HasErrors = hasErrors;
+            }
+
+            public Collection<PSObject> Output { get; }
+
+            public PSDataCollection<ErrorRecord> Errors { get; }
+
+            public bool HasErrors { get; }
+        }
+
+        /// <summary>
+        /// Struct to manage a call that may change the PSModulePath, so that it can be safely reset afterward.
+        /// </summary>
+        /// <remarks>
+        /// If the user manages to set the module path at the same time, using this struct may override that.
+        /// But this happening is less likely than the current issue where the PSModulePath is always reset.
+        /// </remarks>
+        private struct PSModulePathPreserver : IDisposable
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Implement 'IEquatable<T>' in value type 'PSModulePathPreserver'.](https://app.codacy.com/manual/TylerLeonhardt/PowerShellEditorServices/pullRequest?prid=4961943)